### PR TITLE
fix: Window or document access across the code

### PR DESCRIPTION
**Pull request by benjackwhite**
## Changes

Related to https://github.com/PostHog/posthog-js/issues/547 trying to see if we can get the SDK working in service workers nicely.

Direct window/document access is no longer allowed and it is correctly typed now so we have to handle it being undefined

## Checklist
- [ ] Tests for new code (see [advice on the tests we use](https://github.com/PostHog/posthog-js#tiers-of-testing))
- [ ] Accounted for the impact of any changes across different browsers

### CODE REVIEW

**Review by pauldambra — APPROVED** [review]
Nice!

File: src/utils/globals.ts [hunk 1]
```diff
@@ -4,9 +4,11 @@
...
-export const win: Window & typeof globalThis = typeof window !== 'undefined' ? window : ({} as typeof window)
-const navigator = win.navigator || { userAgent: '' }
-const document = win.document || {}
-const userAgent = navigator.userAgent
+// eslint-disable-next-line no-restricted-globals
+export const win: (Window & typeof globalThis) | undefined = typeof window !== 'undefined' ? window : undefined
+const navigator = win?.navigator
+export const document = win?.document
+export const userAgent = navigator?.userAgent
+export const assignableWindow: Window & typeof globalThis & Record<string, any> = win ?? ({} as any)
```
Comment (pauldambra) [inline src/utils/globals.ts:12]: The `Record<string, any>` makes me wonder if we shouldn't make that an explicit type of the things we have added to window... but that can be a follow-up no reason to add more here

**Review by pauldambra — APPROVED** [review]
test pass so 🤞 🚀 :shipit:

File: src/posthog-core.ts [hunk 2]
```diff
@@ -558,7 +557,7 @@ export class PostHog {
 
         // this happens after so a user can call identify in
         // the loaded callback
-        if (this.config.capture_pageview) {
+        if (this.config.capture_pageview && document) {
```
[the remark this answers is not in the document]
Comment (benjackwhite) [reply src/posthog-core.ts:560]: Yep. Which makes sense as there is technically no "page"

**Pull request merged**